stdlib/public/core/DiscontiguousSlice.swift

```diff
@@ -0,0 +1,451 @@
+    var count = 0
+    for element in self {
+      hasher.combine(element)
+      count += 1
+    }
+    hasher.combine(count) // discriminator
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: CustomStringConvertible {
+  public var description: String {
+    _makeCollectionDescription()
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice {
+  /// A position in a `DiscontiguousSlice`.
+  public struct Index {
+    /// The index of the range that contains `base`.
+    internal let _rangeOffset: Int
+    
+    /// The position of this index in the base collection.
+    public let base: Base.Index
+
+    internal init(_rangeOffset: Int, base: Base.Index) {
+      self._rangeOffset = _rangeOffset
+      self.base = base
+    }
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Equatable {
+  public static func == (left: Self, right: Self) -> Bool {
+    left.base == right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Hashable where Base.Index: Hashable {
+  public func hash(into hasher: inout Hasher) {
+    hasher.combine(base)
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Comparable {
+  public static func < (left: Self, right: Self) -> Bool {
+    left.base < right.base
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: CustomStringConvertible {
+  public var description: String {
+    var d = ""
+    debugPrint(base, terminator: "", to: &d)
+    d += " (rangeOffset: \(_rangeOffset))"
+    return d
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice.Index: Sendable where Base.Index: Sendable {}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Sequence {
+  public typealias Element = Base.Element
+  public typealias Iterator = IndexingIterator<Self>
+
+  public func _customContainsEquatableElement(_ element: Element) -> Bool? {
+    _customIndexOfEquatableElement(element).map { $0 != nil }
+  }
+
+  public __consuming func _copyToContiguousArray() -> ContiguousArray<Element> {
+    var result: ContiguousArray<Element> = []
+    for range in subranges.ranges {
+      result.append(contentsOf: base[range])
+    }
+    return result
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+extension DiscontiguousSlice: Collection {
+  public typealias SubSequence = Self
+  public typealias Indices = DefaultIndices<Self>
+  
+  public var startIndex: Index {
+    subranges.isEmpty
+      ? endIndex
+      : Index(_rangeOffset: 0, base: subranges.ranges[0].lowerBound)
+  }
+  
+  public var endIndex: Index {
+    Index(_rangeOffset: subranges.ranges.endIndex, base: _base.endIndex)
+  }
+  
+  public var count: Int {
+    var c = 0
+    for range in subranges.ranges {
+      c += base[range].count
+    }
+    return c
+  }
+
+  public var isEmpty: Bool {
+    subranges.isEmpty
+  }
+
+  public func distance(from start: Index, to end: Index) -> Int {
```

Comment:
We also need to make sure these core Collection APIs are adequately covered by the new tests.

Reply:
The slice tests we have for a normal slice as well as a "complicated" string slice (that involves rounding indices) have pretty good coverage over these